### PR TITLE
Map post popup

### PR DESCRIPTION
**Pull request by TheTexanCodeur**
This pull request implements a map pin popup that enables the user to see more information about a post when he clicks on the map pin. Here are the main features brought with this PR:

1. **Callback function** : The pins now have a callback function that display a popup on top of the map. Nearby and personal post popups display the title and the beginning of the content. The challenge post popups only display the title (keep the "challenge" aspect). 

3. **Details button** : The pin popups have an arrow button that display more detailed information.

- Nearby post : displays the post page
- Personal post : navigate to user profile
- Challenge post : nothing (no arrow button)

5. **Testing** : additional tests are implemented for this new feature. They verify that the new callback function and the map pin popup work as expected. 

Relevant commits for this PR start at 318fdf6

Acceptance Criteria:
- [x] Create the Penpot design of this pop up displaying at least:
   - The title of the post
   - The description of the post
   - A button that redirects to the post page of that particular post.
- [x] Upon clicking a post pin, a pop-up appears 
- [x] Upon clicking on the redirection button within the pop up, the user should redirected to the post page.

### CODE REVIEW

File: lib/viewmodels/map/map_pin_view_model.dart [hunk 1]
```diff
@@ -34,6 +42,13 @@ class MapPinViewModel extends AutoDisposeAsyncNotifier<List<MapPinDetails>> {
     }
   }
 
+  BuildContext? context;
```
Comment (JoachimFavre) [inline lib/viewmodels/map/map_pin_view_model.dart:45]: I don't think this fits here. The `BuildContext` should not be passed in parameter to any function, even less to a view-model (a `BuildContext` is and should stay on the ui layer).

I think this file should be refactored by removing the idea of callback from the pins, but replacing it with parameters such as a `MapPopUpDetails`, which would contain a `title`, a nullable `content`, and a nullable function to call when the user taps on the button (if the function is null, the button is not displayed).

You should be able to reuse most of your code, so I'll still make comments on the code of this file (even if most of it will probably move).

File: lib/viewmodels/map/map_pin_view_model.dart [hunk 2]
```diff
@@ -42,18 +57,66 @@ class MapPinViewModel extends AutoDisposeAsyncNotifier<List<MapPinDetails>> {
...
+                                publicationDate:
+                                    post.data.publicationTime.toDate(),
+                                commentNumber: post.data.commentCount,
+                                voteScore: post.data.voteScore,
+                                distance:
+                                    (GeoFirePoint(position).distanceBetweenInKm(
+                                              geopoint: post.location.geoPoint,
+                                            ) *
+                                            1000)
+                                        .round(),
```
Comment (JoachimFavre) [inline lib/viewmodels/map/map_pin_view_model.dart:109]: There is quite a bit of code repetition between this and `posts_feed_view_model.dart` (lines 85 to 102). I think it would be interesting to add a factory `PostDetails.fromFirestoreData` which would take in parameter the `PostFiresore` post, the `UserFirestore` owner and the `GeoFirePoint` user position.

File: lib/viewmodels/map/map_pin_view_model.dart [hunk 3]
```diff
@@ -88,7 +175,23 @@ class MapPinViewModel extends AutoDisposeAsyncNotifier<List<MapPinDetails>> {
...
     );
 
-    return posts.map(_toMapPinDetails).toList();
+    return posts
+        .map(
+          (post) => _toMapPinDetails(
+            post,
+            callback: context != null
+                ? () => showDialog(
+                      context: context!,
```
Comment (JoachimFavre) [inline lib/viewmodels/map/map_pin_view_model.dart:184]: Clicking on the button for `my posts` brings us on the profile. I'm therefore wondering if clicking on the button on a challenge should bring us to the challenge page. I'm curious to have you opinion on this idea, what do you think of this?

File: lib/views/pages/home/content/map/components/map_pin_pop_up.dart [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    this.title,
+    this.content,
+    this.navigationAction,
+    this.displayButton = false,
+  });
+
+  final String? title;
+  final String? content;
+  final void Function()? navigationAction;
+  final bool displayButton;
```
Comment (JoachimFavre) [inline lib/views/pages/home/content/map/components/map_pin_pop_up.dart:17]: It could be possible that someone cretes this pop up with a navigation action but forgetting to set `displayButton` to true; and it would be possible to have a button that does nothing. I recommend to instead encoode the `displayButton` as whether `navigationAction` is null or not. If this function is null, you don't show the button, if it's not you show it.

File: lib/views/pages/home/content/map/components/map_pin_pop_up.dart [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+  static const arrowButtonKey = Key("arrowButton");
+  const MapPinPopUp({
+    super.key,
+    this.title,
+    this.content,
+    this.navigationAction,
+    this.displayButton = false,
+  });
+
+  final String? title;
```
Comment (JoachimFavre) [inline lib/views/pages/home/content/map/components/map_pin_pop_up.dart:14]: Is it possible to have a pop-up without a title? I feel like this is a bit weird. I think it would be nice to make this parameter non-nullable, and use `required this.title` in the constructor.

File: lib/views/pages/home/content/map/components/map_pin_pop_up.dart [hunk 6]
```diff
@@ -0,0 +1,74 @@
+import "package:flutter/material.dart";
+
+class MapPinPopUp extends StatelessWidget {
```
Comment (JoachimFavre) [inline lib/views/pages/home/content/map/components/map_pin_pop_up.dart:3]: There's a huge amount of code repetition between this and the `profile_info_pop_up.dart`. I think it would be better to make some abstract class that factorises most of the code with some parameter for the `title`, `content?` and `actions`, and make `MapPinPopUp` and `ProfileInfoPopUp` inherit from it.

Moreover, this class is missing class documentation ;)

File: test/viewmodels/map/map_pin_view_model_integration_test.dart [hunk 7]
```diff
@@ -118,6 +141,60 @@ void main() {
 
         expect(pinListTuple, unorderedEquals(expectedPinsTuple));
       });
+
+      testWidgets("callback function is created as expected", (tester) async {
```
Comment (JoachimFavre) [inline test/viewmodels/map/map_pin_view_model_integration_test.dart:145]: This test does not fit in this file. `testWidgets` tests the UI, when here you want to test a view model. I think this test should be moved to one of the `views/pages/map` test suite.

File: test/viewmodels/map/map_pin_view_model_integration_test.dart [hunk 8]
```diff
@@ -19,22 +27,37 @@ void main() {
...
+          .thenAnswer((_) async => testingUserFirestore);
+
       container = ProviderContainer(
         overrides: [
-          firestoreProvider.overrideWithValue(fakeFireStore),
           geolocationServiceProvider.overrideWithValue(geolocationService),
+          loggedInUserIdProvider.overrideWithValue(testingUserFirestoreId),
+          firestoreProvider.overrideWithValue(fakeFireStore),
+          userRepositoryServiceProvider
+              .overrideWithValue(userRepositoryService),
```
Comment (JoachimFavre) [inline test/viewmodels/map/map_pin_view_model_integration_test.dart:60]: By humble curiosity, why is this necessary when you already have `loggedInUserIdProvide.overrideWithValue(testingUserFirestore)`?

**Review by JoachimFavre — CHANGES_REQUESTED** [review]
You have written interesting code. However, I think it has some issues, the most important one being its separation of layers (the difference between UI and view-model being not respected enough in my opinion); I'll let you read my comments below for more information. I moreover have two more general comments:
1) You created a `views/pages/home/map/component` package, shouldn't `views/pages/home/map/map_screen.dart` go in it too?
2) You test the callback function of only one of the three possibilities. The callback for "my posts" and the absence of callback for challenges (or its callback if you follow one of my suggestions to add it one) should be tested too in my opinion.

If you have any hesitation on something or need any help, do not hesitate to ask me questions. I want to do my best to help you finish solving those problems as soon as possible! :)

File: lib/viewmodels/map/map_pin_view_model.dart (hunk 2 above)
Comment (TheTexanCodeur) [reply lib/viewmodels/map/map_pin_view_model.dart]: Fixed with [5223934](https://github.com/ProximaEPFL/proxima/pull/335/commits/5223934e9ef70f7e76d529b65a3fd553cfbe35ca)

File: lib/viewmodels/map/map_pin_view_model.dart (hunk 3 above)
Comment (TheTexanCodeur) [reply lib/viewmodels/map/map_pin_view_model.dart:172]: It was also my initial idea, but to be honest I didn't find a way to navigate to this page. Triggering a navigation of the bottom navigation bar revealed itself harder than planned and started to fall out of the scope of the PR.

File: test/mocks/providers/provider_map_page.dart [hunk 9]
```diff
@@ -46,3 +52,20 @@ ProviderScope newMapPageWithPins(
     child: mapPage,
   );
 }
+
+ProviderScope newMapPageWithPinsNoOverride(
```
[the remark this answers is not in the document]
Comment (TheTexanCodeur) [reply test/mocks/providers/provider_map_page.dart]: Fixed with [eaeac6f](https://github.com/ProximaEPFL/proxima/pull/335/commits/eaeac6feee26414b136ebb573503c1e0263c9403)

File: lib/viewmodels/map/map_pin_view_model.dart (hunk 1 above)
Comment (TheTexanCodeur) [reply lib/viewmodels/map/map_pin_view_model.dart]: fixed with [994d03c](https://github.com/ProximaEPFL/proxima/pull/335/commits/994d03c1af074c06ac3ec0bd4f9b9c8c5ff4534d)

File: lib/views/pages/home/content/map/components/map_pin_pop_up.dart (hunk 4 above)
Comment (TheTexanCodeur) [reply lib/views/pages/home/content/map/components/map_pin_pop_up.dart]: Fixed with [994d03c](https://github.com/ProximaEPFL/proxima/pull/335/commits/994d03c1af074c06ac3ec0bd4f9b9c8c5ff4534d)

File: lib/views/pages/home/content/map/components/map_pin_pop_up.dart (hunk 5 above)
Comment (TheTexanCodeur) [reply lib/views/pages/home/content/map/components/map_pin_pop_up.dart]: fixed with [791bcf4](https://github.com/ProximaEPFL/proxima/pull/335/commits/791bcf45e1f74336507c3693a62c575ce7bb3fe2)

File: lib/views/pages/home/content/map/components/map_pin_pop_up.dart (hunk 6 above)
Comment (TheTexanCodeur) [reply lib/views/pages/home/content/map/components/map_pin_pop_up.dart:4]: Fixed with [3845991](https://github.com/ProximaEPFL/proxima/pull/335/commits/3845991001207e3aba839c1890e902ea4af3ae33)

File: lib/viewmodels/map/map_pin_view_model.dart (hunk 3 above)
Comment (JoachimFavre) [reply lib/viewmodels/map/map_pin_view_model.dart]: Yes ok this makes a lot of sense, I understand better. This is fine as you did! :)

File: test/viewmodels/map/map_pin_view_model_integration_test.dart [hunk 10]
```diff
@@ -118,6 +141,60 @@ void main() {
...
+        await tester.pumpAndSettle();
+
+        final pinList = await container.read(mapPinViewModelProvider.future);
+        await tester.pumpAndSettle();
+
+        expect(pinList, isNotEmpty);
+
+        //check that the callback function of the pin shows a dialog
+        final pin = pinList.first;
+        pin.callbackFunction();
```
[the remark this answers is not in the document]
Comment (TheTexanCodeur) [reply test/viewmodels/map/map_pin_view_model_integration_test.dart]: No we can not access the Marker of the map sadly :(

File: test/viewmodels/map/map_pin_view_model_integration_test.dart (hunk 8 above)
Comment (TheTexanCodeur) [reply test/viewmodels/map/map_pin_view_model_integration_test.dart:58]: because the map pin viewmodel also needs the user info and not only the user ID

Comment (JoachimFavre) [reply test/viewmodels/map/map_pin_view_model_integration_test.dart:60]: Could `loggedInUserIdProvide.overrideWithValue(testingUserFirestore)` be removed then?

File: test/viewmodels/map/map_pin_view_model_integration_test.dart [hunk 11]
```diff
@@ -118,6 +141,60 @@ void main() {
...
+
+        //click on the button in the popup
+        final arrowButton = find.byKey(MapPinPopUp.arrowButtonKey);
+        expect(arrowButton, findsOneWidget);
+
+        await tester.tap(arrowButton);
+
+        await tester.pumpAndSettle();
+        //check that we have a post overview page
+        expect(find.byType(PostPage), findsOneWidget);
```
[the remark this answers is not in the document]
Comment (TheTexanCodeur) [reply test/viewmodels/map/map_pin_view_model_integration_test.dart:203]: Fixed with [911786d](https://github.com/ProximaEPFL/proxima/pull/335/commits/911786d7624898fc7411741b6a5058de6057895d)

File: test/viewmodels/map/map_pin_view_model_integration_test.dart (hunk 7 above)
Comment (TheTexanCodeur) [reply test/viewmodels/map/map_pin_view_model_integration_test.dart]: Fixed with [5179bf4](https://github.com/ProximaEPFL/proxima/pull/335/commits/5179bf4d69d67e32eea422348a488a327aa482dd)

File: test/viewmodels/map/map_pin_view_model_integration_test.dart [hunk 12]
```diff
@@ -118,6 +141,60 @@ void main() {
 
         expect(pinListTuple, unorderedEquals(expectedPinsTuple));
       });
+
+      testWidgets("callback function is created as expected", (tester) async {
+        // Generate 10 random positions around the user but in range to be able to see them
```
[the remark this answers is not in the document]
Comment (TheTexanCodeur) [reply test/viewmodels/map/map_pin_view_model_integration_test.dart]: Fixed with [5179bf4](https://github.com/ProximaEPFL/proxima/pull/335/commits/5179bf4d69d67e32eea422348a488a327aa482dd)

File: test/viewmodels/map/map_pin_view_model_integration_test.dart (hunk 8 above)
Comment (TheTexanCodeur) [reply test/viewmodels/map/map_pin_view_model_integration_test.dart:48]: No because it need this id to get the rigth user info in the user repo

File: test/viewmodels/map/map_pin_view_model_integration_test.dart [hunk 13]
```diff
@@ -150,17 +227,24 @@ void main() {
...
             mockGeolocator.getPositionStream(
               locationSettings: geolocationService.locationSettings,
             ),
           ).thenAnswer(
-            (_) => Stream.fromIterable(
-              postPositions.map(
-                (geoPoint) =>
-                    getSimplePosition(geoPoint.latitude, geoPoint.longitude),
-              ),
+            //We need to introduce delay in this test
```
[the remark this answers is not in the document]
Comment (TheTexanCodeur) [reply test/viewmodels/map/map_pin_view_model_integration_test.dart:169]: I added more doc with [ca669a6](https://github.com/ProximaEPFL/proxima/pull/335/commits/ca669a60a0bcb18810293c8ebba64fca995627d4)

File: test/mocks/overrides/override_pin_view_model.dart [hunk 14]
```diff
@@ -11,6 +12,7 @@ class MockPinViewModel extends AutoDisposeAsyncNotifier<List<MapPinDetails>>
 
   MockPinViewModel({
     Future<List<MapPinDetails>> Function()? build,
+    Function(BuildContext)? setContext,
```
Comment (JoachimFavre) [inline test/mocks/overrides/override_pin_view_model.dart:15]: You no longer need this.

**Review by JoachimFavre — COMMENTED** [review]
I just have some small comments, but it's very close to finishing the PR! We're almost there! :)

File: test/viewmodels/map/map_pin_view_model_integration_test.dart (hunk 8 above)
Comment (JoachimFavre) [reply test/viewmodels/map/map_pin_view_model_integration_test.dart:50]: Makes sense, thank you ! 

In fact, after the test refactor, this was no longer even applicable.

File: test/viewmodels/map/map_pin_view_model_integration_test.dart (hunk 13 above)
[the remark this answers is not in the document]
Comment (JoachimFavre) [reply test/viewmodels/map/map_pin_view_model_integration_test.dart:171]: Ok, I understand this better. This is quite a weird test haha, and it must have been hard for you to find how to fix it.

File: test/mocks/overrides/override_pin_view_model.dart (hunk 14 above)
Comment (TheTexanCodeur) [reply test/mocks/overrides/override_pin_view_model.dart]: Fixed with [82d65b9](https://github.com/ProximaEPFL/proxima/pull/335/commits/82d65b9eff2d4e6a69793b06a208dede64fa5cd8)

File: test/mocks/providers/provider_map_page.dart (hunk 9 above)
[the remark this answers is not in the document]
Comment (JoachimFavre) [reply test/mocks/providers/provider_map_page.dart]: In fact, it appeared this provider was no longer used after the test refactor, and it was removed in 6b340a515ac5a9a7cece411ec64d61db66a7a5bf

File: test/viewmodels/map/map_pin_view_model_integration_test.dart (hunk 11 above)
[the remark this answers is not in the document]
Comment (JoachimFavre) [reply test/viewmodels/map/map_pin_view_model_integration_test.dart]: Thank you! :)

File: lib/models/ui/post_details.dart [hunk 15]
```diff
@@ -63,4 +66,29 @@ class PostDetails {
       isChallenge,
     );
   }
+
+  factory PostDetails.fromFirestoreData(
```
[the remark this answers is not in the document]
Comment (JoachimFavre) [reply lib/models/ui/post_details.dart:75]: Fixed in 08d1bbb055d501e207f0fd1274dc6d9254ae056f. Thanks! :)

**Review by JoachimFavre — APPROVED** [review]
LGTM now! Good job! 

These were big refactors, and you handled them like a pro! Nice work :)
